src/IssueViz.Security/OpenInIDE/Api/IOpenInIDEStateValidator.cs

```diff
@@ -41,4 +46,103 @@ internal interface IOpenInIDEStateValidator
         /// </remarks>
         bool CanHandleOpenInIDERequest(Uri serverUri, string projectKey, string organizationKey);
     }
+
+    [Export(typeof(IOpenInIDEStateValidator))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    internal class OpenInIdeStateValidator : IOpenInIDEStateValidator
+    {
+        private readonly IInfoBarManager infoBarManager;
+        private readonly IConfigurationProvider configurationProvider;
+        private readonly ILogger logger;
+        private IInfoBar currentInfoBar;
+
+        [ImportingConstructor]
+        public OpenInIdeStateValidator(IInfoBarManager infoBarManager, IConfigurationProvider configurationProvider, ILogger logger)
+        {
+            this.infoBarManager = infoBarManager;
+            this.configurationProvider = configurationProvider;
+            this.logger = logger;
+        }
+
+        public bool CanHandleOpenInIDERequest(Uri serverUri, string projectKey, string organizationKey)
+        {
+            RemoveExistingInfoBar();
+
+            var failureMessage = GetFailureMessage(serverUri, projectKey, organizationKey);
+
+            if (string.IsNullOrEmpty(failureMessage))
+            {
+                return true;
+            }
+
+            AddInfoBar(failureMessage);
+            logger.WriteLine(failureMessage);
+
+            return false;
+        }
+
+        private string GetFailureMessage(Uri serverUri, string projectKey, string organizationKey)
+        {
+            var configuration = configurationProvider.GetConfiguration();
+            string reason = null;
+
+            if (configuration.Mode == SonarLintMode.Standalone)
+            {
+                reason = OpenInIDEResources.Infobar_InvalidStateReason_NotInConnectedMode;
+            }
+            else if (!configuration.Project.ServerUri.Equals(serverUri))
+            {
+                reason = string.Format(OpenInIDEResources.Infobar_InvalidStateReason_WrongServer, configuration.Project.ServerUri);
+            }
+            else if (!string.IsNullOrEmpty(organizationKey) &&
+                (configuration.Project.Organization == null ||
+                 !organizationKey.Equals(configuration.Project.Organization.Key, StringComparison.OrdinalIgnoreCase)))
+            {
+                reason = string.Format(OpenInIDEResources.Infobar_InvalidStateReason_WrongOrganization, configuration.Project.Organization?.Key);
```

Comment:
I decided to use the same message for incorrect and null organization. Null organization is not supposed to happen, since you would have to be bound to the correct server (=sonarcloud) but with no organization, or we would have to get a request from SQ with SQ server and not-null organization. Both cases are not supposed to happen, so I am handling them but the user-message would be partial (`you must connect to the right organization. Currently connected to: `)